cpp/unreal_plugins/SimulationController/Source/SimulationController/WheeledVehicleAgent.cpp

```diff
@@ -0,0 +1,310 @@
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, ArrayDesc> WheeledVehicleAgent::getStepInfoSpace() const
+{
+    return {};
+}
+
+void WheeledVehicleAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& action)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "apply_wheel_torques")) {
+        std::vector<double> component_data = Std::reinterpretAs<double>(action.at("apply_wheel_torques"));
+        wheeled_vehicle_pawn_->setDriveTorques(Eigen::Vector4d(
+            component_data.at(0), component_data.at(1), component_data.at(2), component_data.at(3)));
+        wheeled_vehicle_pawn_->setBrakeTorques(Eigen::Vector4d(0.0f, 0.0f, 0.0f, 0.0f));
+    }
+
+    if (Std::contains(action_components, "set_position_xyz_centimeters")) {
+        std::vector<double> component_data = Std::reinterpretAs<double>(action.at("set_position_xyz_centimeters"));
+        FVector location(component_data.at(0), component_data.at(1), component_data.at(2));
+        bool sweep = false;
+        FHitResult* hit_result = nullptr;
+        wheeled_vehicle_pawn_->SetActorLocation(location, sweep, hit_result, ETeleportType::TeleportPhysics);
+        wheeled_vehicle_pawn_->setBrakeTorques(Eigen::Vector4d(1000.0f, 1000.0f, 1000.0f, 1000.0f)); // TODO: get value from the config system
+    }
+
+    if (Std::contains(action_components, "set_orientation_pyr_radians")) {
+        std::vector<double> component_data = Std::reinterpretAs<double>(action.at("set_orientation_pyr_radians"));
+        FRotator rotation(
+            FMath::RadiansToDegrees(component_data.at(0)),
+            FMath::RadiansToDegrees(component_data.at(1)),
+            FMath::RadiansToDegrees(component_data.at(2)));
+        wheeled_vehicle_pawn_->SetActorRotation(rotation, ETeleportType::TeleportPhysics);
+        wheeled_vehicle_pawn_->setBrakeTorques(Eigen::Vector4d(1000.0f, 1000.0f, 1000.0f, 1000.0f)); // TODO: get value from the config system
+    }
+}
+
+std::map<std::string, std::vector<uint8_t>> WheeledVehicleAgent::getObservation() const
+{
+    std::map<std::string, std::vector<uint8_t>> observation;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "state_data")) {
+        FVector location = wheeled_vehicle_pawn_->GetActorLocation();
+        FRotator rotation = wheeled_vehicle_pawn_->GetActorRotation();
+        observation["state_data"] = Std::reinterpretAs<uint8_t>(std::vector<double>{
+            location.X,
+            location.Y,
+            location.Z,
+            FMath::DegreesToRadians(rotation.Pitch),
```

Comment:
Don't try to reformat the data coming out of Unreal.

Reply:
Agreed, updated this.